Add fl-with-flower post (#919)
huggingface/blog: fl-with-flower.md

---
title: "Federated Learning using Hugging Face and Flower" 
thumbnail: /blog/assets/fl-with-flower/thumbnail.png
authors:
- user: charlesbvll
  guest: true
---

# Federated Learning using Hugging Face and Flower

<!-- {blog_metadata} -->
<!-- {authors} -->

This tutorial will show how to leverage Hugging Face to federate the training of language models over multiple clients using [Flower](https://flower.dev/). More specifically, we will fine-tune a pre-trained Transformer model (distilBERT) for sequence classification over a dataset of IMDB ratings. The end goal is to detect if a movie rating is positive or negative.

## Standard Hugging Face workflow

### Handling the data

To fetch the IMDB dataset, we will use Hugging Face's `datasets` library. We then need to tokenize the data and create `PyTorch` dataloaders, this is all done in the `load_data` function:

```python
def load_data():
    """Load IMDB data (training and eval)"""
    raw_datasets = load_dataset("imdb")
    raw_datasets = raw_datasets.shuffle(seed=42)

    # remove unnecessary data split
    del raw_datasets["unsupervised"]

>     tokenizer = AutoTokenizer.from_pretrained(CHECKPOINT)

Reviewer: Would be good to have `CHECKPOINT` defined

To create the model itself, we will just load the pre-trained distillBERT model using the Hugging Face’s `AutoModelForSequenceClassification` :

```python
net = AutoModelForSequenceClassification.from_pretrained(
        "distilbert-base-uncased", num_labels=2
    ).to(torch.device("cuda:0" if torch.cuda.is_available() else "cpu"))
```

## Federating the example

The idea behind Federated Learning is to train a model between multiple clients and a server, without having to share any data. This is done by letting each client train the model locally on its data and send its parameters back to the server which then aggregates all the clients’ parameters together using a predefined strategy. This process is made very simple by using the [Flower](https://github.com/adap/flower) framework. If you want a more complete overview, be sure to check out this guide: [What is Federated Learning?](https://flower.dev/docs/tutorial/Flower-0-What-is-FL.html)

### Creating the IMDBClient

To federate our example to multiple clients, we first need to write our Flower client class (inheriting from `flwr.client.NumPyClient`). This is very easy, as our model is a standard `PyTorch` model:

```python

> class IMDBClient(fl.client.NumPyClient):

Reviewer: here and in other cells I would add imports corresponding to the cell

        def get_parameters(self, config):
            return [val.cpu().numpy() for _, val in net.state_dict().items()]

        def set_parameters(self, parameters):
            params_dict = zip(net.state_dict().keys(), parameters)
            state_dict = OrderedDict({k: torch.Tensor(v) for k, v in params_dict})
            net.load_state_dict(state_dict, strict=True)

        def fit(self, parameters, config):
            self.set_parameters(parameters)
            print("Training Started...")
            train(net, trainloader, epochs=1)
            print("Training Finished.")
            return self.get_parameters(config={}), len(trainloader), {}

        def evaluate(self, parameters, config):
            self.set_parameters(parameters)
            loss, accuracy = test(net, testloader)
            return float(loss), len(testloader), {"accuracy": float(accuracy)}
```

> The `get_parameters` function allows the server to get the client's parameters, inversely, the `set_parameters` function allows the server to send its parameters to the client. Finally the `fit` function is to train the model locally for the client and the `evaluate` function is to test the model locally and return the relevant metrics. 

Reviewer: ```suggestion
The `get_parameters` function lets the server get the client's parameters. Inversely, the `set_parameters` function allows the server to send its parameters to the client. Finally, the `fit` function trains the model locally for the client, and the `evaluate` function tests the model locally and returns the relevant metrics. 
```

We can now start client instances using:

```python
flwr.client.start_numpy_client(server_address="127.0.0.1:8080", 
															 client=IMDBClient())
```

### Starting the server

> Once we have a way to instantiate clients, we need to create our server in order to aggregate the results. Using Flower, this can be done very easily, by first choosing a strategy and then using the `flwr.server.start_server` function:

Reviewer: ```suggestion
Once we have a way to instantiate clients, we need to create our server in order to aggregate the results. Using Flower, this can be done very easily by first choosing a strategy and then using the `flwr.server.start_server` function:
```

---
title: "Federated Learning using Hugging Face and Flower" 
thumbnail: /blog/assets/fl-with-flower/thumbnail.png
authors:
- user: charlesbvll
  guest: true
---

# Federated Learning using Hugging Face and Flower

<!-- {blog_metadata} -->
<!-- {authors} -->

<a target="_blank" href="https://colab.research.google.com/github/huggingface/blog/blob/main/notebooks/fl-with-flower.ipynb">
    <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>

This tutorial will show how to leverage Hugging Face to federate the training of language models over multiple clients using [Flower](https://flower.dev/). More specifically, we will fine-tune a pre-trained Transformer model (distilBERT) for sequence classification over a dataset of IMDB ratings. The end goal is to detect if a movie rating is positive or negative.

## Dependencies

To follow along this tutorial you will need to install the following packages: `datasets`, `flwr`, `torch`, and `transformers`. This can be done using `pip`:

```sh

> pip install datasets flwr torch transformers

Reviewer: Maybe mention to use `flwr["simulation"]` if in colab

```

## Standard Hugging Face workflow

### Handling the data

To fetch the IMDB dataset, we will use Hugging Face's `datasets` library. We then need to tokenize the data and create `PyTorch` dataloaders, this is all done in the `load_data` function:

```python
import random

import torch
from datasets import load_dataset
from torch.utils.data import DataLoader
from transformers import AutoTokenizer, DataCollatorWithPadding

DEVICE = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
CHECKPOINT = "distilbert-base-uncased"

def load_data():
    """Load IMDB data (training and eval)"""
    raw_datasets = load_dataset("imdb")
    raw_datasets = raw_datasets.shuffle(seed=42)

    # remove unnecessary data split
    del raw_datasets["unsupervised"]

    tokenizer = AutoTokenizer.from_pretrained(CHECKPOINT)

    def tokenize_function(examples):
        return tokenizer(examples["text"], truncation=True)

    # random 100 samples
    population = random.sample(range(len(raw_datasets["train"])), 100)

    tokenized_datasets = raw_datasets.map(tokenize_function, batched=True)

>     tokenized_datasets["train"] = tokenized_datasets["train"].select(population)

Reviewer: Maybe add a comment on why we are selecting just a couple of samples

    tokenized_datasets["test"] = tokenized_datasets["test"].select(population)

    tokenized_datasets = tokenized_datasets.remove_columns("text")
    tokenized_datasets = tokenized_datasets.rename_column("label", "labels")

    data_collator = DataCollatorWithPadding(tokenizer=tokenizer)
    trainloader = DataLoader(
        tokenized_datasets["train"],
        shuffle=True,
        batch_size=32,
        collate_fn=data_collator,
    )

    testloader = DataLoader(
        tokenized_datasets["test"], batch_size=32, collate_fn=data_collator
    )

    return trainloader, testloader
```

### Training and testing the model

Once we have a way of creating our trainloader and testloader, we can take care of the training and testing. This is very similar to any `PyTorch` training or testing loop:

```python

> import torch

Reviewer: No need to import torch in every cell, just the first time it's used

> from datasets import load_metric

Reviewer 2: metrics are deprecated in `datasets`, you can use `evaluate` instead:

```suggestion
from evaluate import load as load_metric
```

        def evaluate(self, parameters, config):
            self.set_parameters(parameters)
            loss, accuracy = test(net, testloader)
            return float(loss), len(testloader), {"accuracy": float(accuracy)}
```

The `get_parameters` function lets the server get the client's parameters. Inversely, the `set_parameters` function allows the server to send its parameters to the client. Finally, the `fit` function trains the model locally for the client, and the `evaluate` function tests the model locally and returns the relevant metrics. 

We can now start client instances using:

```python
import flwr as fl

fl.client.start_numpy_client(server_address="127.0.0.1:8080", 
															 client=IMDBClient())
```

### Starting the server

Now that we have a way to instantiate clients, we need to create our server in order to aggregate the results. Using Flower, this can be done very easily by first choosing a strategy (here, we are using `FedAvg`, which will define the global weights as the average of all the clients' weights at each round) and then using the `flwr.server.start_server` function:

```python

> import flwr as fl

Reviewer: If this is already imported once above, no need to import every single time, just the first time

def weighted_average(metrics):
    accuracies = [num_examples * m["accuracy"] for num_examples, m in metrics]
    losses = [num_examples * m["loss"] for num_examples, m in metrics]
    examples = [num_examples for num_examples, _ in metrics]
    return {"accuracy": sum(accuracies) / sum(examples), "loss": sum(losses) / sum(examples)}

# Define strategy
strategy = fl.server.strategy.FedAvg(
    fraction_fit=1.0,
    fraction_evaluate=1.0,
    evaluate_metrics_aggregation_fn=weighted_average,
)

# Start server
fl.server.start_server(
    server_address="0.0.0.0:8080",
    config=fl.server.ServerConfig(num_rounds=3),
    strategy=strategy,
)
```

The `weighted_average` function is there to provide a way to aggregate the metrics distributed amongst the clients (basically this allows us to display a nice average accuracy and loss for every round).

## Putting everything together

If you want to check out everything put together, you should check out the code example we wrote for the Flower repo: [https://github.com/adap/flower/tree/main/examples/quickstart_huggingface](https://github.com/adap/flower/tree/main/examples/quickstart_huggingface). 

> A notebook is also available here: [https://colab.research.google.com/github/huggingface/blog/blob/main/notebooks/fl-with-flower.ipynb](https://colab.research.google.com/github/huggingface/blog/blob/main/notebooks/fl-with-flower.ipynb) but instead of running on multiple separate clients it utilizes the simulation functionality of Flower in order to emulate a federated setting inside Google Colab (this also means that instead of calling `start_server` we will call `start_simulation`, and that a few other modifications are needed).

Reviewer: Would maybe put this closer to the beginning
